1-Draft/RFCXXXX-Umask-and-Ulimit.md

```diff
@@ -0,0 +1,276 @@
+---
+RFC: XXXX
+Author: Jim Truher
+Status: Draft
+Area: Strict Mode
+Comments Due: 7/1/2019
+---
+
+# Umask and Ulimit
+
+Unix shells have a number of built in commands which are used to configure the shell's behavior.
+This RFC proposes two new areas of control for PowerShell, namely, `umask` and `ulimit`
+
+## UMASK
+
+On Linux and MacOS systems when a file is created, it is created with a specific set of permissions.
+This is known as _file mode creation mask_.
+These permissions are set via the system api `umask` and are usually altered by the user with the `umask` command.
+In Unix shells, the `umask` command is built in to the shell, because the new value exists only for as long as the process that started it and any child process.
+Most systems also include `/usr/bin/umask` which is a stand-alone command which may be used to _retrieve_ the value of the mask.
+Using `/usr/bin/umask` can generate confusion as some may believe that setting the mask with `/usr/bin/umask` will persist this setting, but this is not the case as when the invocation of `/usr/bin/umask` concludes, the previous mask will be back in effect.
+For PowerShell customers, this becomes problematic as there is no easy way to invoke a persisted `umask` setting as there is no built-in command to change the `umask` of the PowerShell process.
+
+### Two new cmdlets
+
+This section of the RFC proposes two new cmdlets:
+
+* Get-Umask
+* Set-Umask
+
+These cmdlets (and associated engine support) will be available only on MacOS and Linux.
+`Get-Umask` retrieves the current process umask setting and `Set-Umask` sets the current process umask setting.
+For a good discussion of umask and its use, please visit: https://www.cyberciti.biz/tips/understanding-linux-unix-umask-value-usage.html.
+
+#### Get-Umask Syntax
+
+```powershell
+NAME
+    Get-Umask
+
+SYNTAX
+    Get-Umask [<CommonParameters>]
+
+ALIASES
+    None
+```
+
+#### Get-Umask Examples
+
+When running `Get-Umask` an object will be returned which contains the traditional umask string, the Symbolic representation of the mask, and an integer representation as follows:
+
+```powershell
+PS> get-umask
+
+Umask Symbolic        AsDecimal
+----- --------        ---------
+0022  u=rwx,g=rx,o=rx        18
+```
+
+When used in a string context, the traditional view of umask shall be presented:
+
+```powershell
+PS> "$(get-umask)"
+0022
+```
+
+#### Set-Umask Syntax
+
+```powershell
+NAME
+    Set-Umask
+
+SYNTAX
+    Set-Umask [-Umask] <string> [<CommonParameters>]
+    Set-Umask -Symbolic <string[]> [<CommonParameters>]
+
+ALIASES
+    None
+```
+
+Two parameter sets for `Set-Umask` will be present:
+
+* Umask
+  * A string which represents the umask as an octal number
+* Symbolic
+  * a string which represents the umask as a comma separated group indicating the permissions of the file owner, group, and other.
+
```

Comment:
I think we should discuss what the allowed strings are for each parameter and what the behaviours are when bad strings are given.

A few cases arise in my mind because of the very open spectrum of possible string inputs (vs ints for example):
- `'002'` vs `'2'`
- Giving the argument as a decimal int like `761`, coercing that to the string `'762'` and actually ending up with the decimal permissions `498`. The problem is that a literal decimal will succeed because it gets stringified, but if anyone expects an int to work and use a variable to do arithmetic on it, strange things will happen because of the base change.
- A generic bad string like `'banana'`
- Partial formations of the symbolic string like `'o=` or `'x='` or 'u=b'

Essentially I'm thinking that passing strings allows a whole bunch of inputs that aren't valid and we're using them in both cases with a special parser for a sublanguage. I'm not saying we should definitely be using a different type of input instead, but I think we should discuss and define:
- the language of strings expected by each parameter
- what the behaviours are when a badly formed string input is given
- what the alternative implementations of types for parameters might look like

Another thought that just struct me is that it's quite possible for us to define a new type (a struct perhaps) that will coerce from an int or a string equally, or can be constructed more explicitly.

Reply:
@rjmholt 

Given how unhelpful it is to conceive of and formulate the value as a _decimal_, I don't think we need to worry about decimal input.

Good point about malformed strings; as a data point, Bash's `ulimit` quietly ignores empty values such as `u=` and accepts multiple values such as `u=rr`.

On the flip side, it balks at whitespace and invalid characters, on either side of the `=`, including the uppercase variants of `u`, `o`, and `g`.

"Balking" means that an error is emitted to stderr, and the exit code is set to 1.

In the world of PowerShell that probably corresponds to a statement-terminating error (you're trying to set a single value, and the input is invalid).

If we implement the separate `-User`, `-Group` and `-Other` parameters (which makes sense to me), you implicitly get a structure of sorts, via splatting, which perhaps is enough:

```powershell
$umask = @{
  User = 'rw'
  Group = 'r'
  Other = 'r'
}
Set-Umask @umask
```

Just to spell it out: The separate parameters should have set-a-value-for-that-principal-only logic, not default-the-unspecified-values to-`0`-logic, the way the symbolic value does it (but not the octal one).